### PR TITLE
8253420: Refactor HeapRegionManager::find_highest_free

### PR DESCRIPTION
**Pull request by albertnetymk**
Using for-loop to make the number of iterations more explicit. Direct backward iteration, `for (uint curr = reserved_length() - 1; curr >= 0; curr--)` doesn't work due to underflow of `uint` type. Therefore, I went for current approach.

Test: hotspot_gc

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253420](https://bugs.openjdk.java.net/browse/JDK-8253420): Refactor HeapRegionManager::find_highest_free

### Reviewers
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**)
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2193/head:pull/2193`
`$ git checkout pull/2193`

### CODE REVIEW

File: src/hotspot/share/gc/g1/heapRegionManager.cpp [hunk 1]
```diff
@@ -537,8 +537,8 @@ uint HeapRegionManager::find_highest_free(bool* expanded) {
   // Loop downwards from the highest region index, looking for an
   // entry which is either free or not yet committed.  If not yet
   // committed, expand at that index.
-  uint curr = reserved_length() - 1;
-  while (true) {
+  for (uint i = 0; i < reserved_length(); ++i) {
+    uint curr = reserved_length() - 1 - i;
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/heapRegionManager.cpp:541]: Maybe this instead?
```
for (uint curr = reserved_length(); curr-- > 0; ) {
```

Comment (albertnetymk) [reply src/hotspot/share/gc/g1/heapRegionManager.cpp:541]: I would prefer not having side effect in the condition. At first glance, it's not obvious how many iteration the loop entails, `length` or `length - 1`?

Comment (kstefanj) [reply src/hotspot/share/gc/g1/heapRegionManager.cpp:541]: Avoiding side effects is normally good, but in this case I think it actually make the whole intent of the code clearer. We could add to the comment that we loop backwards through all reserved regions to make it clear what the bound is.

Comment (albertnetymk) [reply src/hotspot/share/gc/g1/heapRegionManager.cpp]: Updated as suggested, since you both think it's better.

**Review by kstefanj — APPROVED** [review]

**Review by kimbarrett — APPROVED** [review]

**Comment by albertnetymk** [comment]
Thank you for the reviews.

/integrate

**Pull request closed without merging**